### PR TITLE
Article: Write a SQL Optimizer using Egg

### PR DESCRIPTION
**Pull request by wangrunji0408**
Sorry for the delay. This article was originally written in Chinese and then translated into English by ChatGPT. This PR contains both version.

### CODE REVIEW

File: content/issue-2/write-a-sql-optimizer-using-egg.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+Firstly, we simplify expressions in the plan nodes.
+
+For example, the multiplication-to-shift rule mentioned at the beginning can be described in Egg using the [`rewrite`](https://docs.rs/egg/0.9.3/egg/macro.rewrite.html) macro as follows:
+
+```rust
+rewrite!("mul2-to-shl1"; "(* ?a 2)" => "(<< ?a 1)")
+```
+
+The strings on either side of `=>` ****represent the expression patterns before and after the transformation, where `?a` represents a variable that can be replaced with any expression. When the runner finds an expression that matches this pattern, it generates a new expression according to the pattern on the right and unions it with the original node to form an equivalence class. By the way, `=>` can also be replaced with `<=>` to indicate a bidirectional transformation.
```
Comment (Folyd) [inline content/issue-2/write-a-sql-optimizer-using-egg.md:127]: What does `****represent` mean?

File: content/issue-2/write-a-sql-optimizer-using-egg.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+In this article, we will briefly introduce the implementation process of this project. You can find the relevant code in the [SQL Optimizer Labs](https://github.com/risinglightdb/sql-optimizer-labs), or learn about its application in the [RisingLight](https://github.com/risinglightdb/risinglight) project.
+
+```urlpreview
+https://github.com/risinglightdb/sql-optimizer-labs
+```
+
+# The Egg Library
+
+
+Egg is a program optimizer framework written in Rust. Its core technology is based on a method called Equality Saturation. The idea behind it is to gradually rewrite expressions to find all equivalent forms and then identify the optimal solution among them. During this process, Egg uses the e-graph data structure to efficiently query and maintain equivalence classes at runtime, reducing the time and space costs of program optimization.
```
Comment (pacak) [inline content/issue-2/write-a-sql-optimizer-using-egg.md:14]:
```suggestion
Egg is a program optimizer framework written in Rust. Its core technology is based on a method called Equality Saturation. The idea behind it is to gradually rewrite expressions to find all equivalent forms and then identify the optimal solution among them. During this process, Egg uses the [e-graph](https://en.wikipedia.org/wiki/E-graph) data structure to efficiently query and maintain equivalence classes at runtime, reducing the time and space costs of program optimization.
```

File: content/issue-2/write-a-sql-optimizer-using-egg.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+![](/static/issue-2/egg/egraph.png)
+
+Each node in the diagram is an e-node, which can represent a variable, a constant, or an operation. Multiple e-nodes can form an e-class, which represents a group of equivalent nodes with the same semantics that can be replaced by each other. The child nodes of each e-node are e-classes, which together form an e-graph. Therefore, this data structure can represent a large number of possible combinations with compact space.
+
+The e-graph also allows for dynamic insertion of e-nodes and merging of e-classes, which enables expression rewriting. The following diagram illustrates the process of inserting a new expression `a << 1` into the graph and merging it with `a * 2`.
+
+![](/static/issue-2/egg/rewrite.png)
+
+Egg supports user-defined rules using Lisp expressions. For example, the rule shown in the above figure can be expressed as `(* ?a 2) => (<< ?a 1)`. More complex rules can also be written in Rust if needed. This makes Egg highly flexible and extensible. Developers can quickly implement an optimizer for their own language based on Egg. The following sections of this article demonstrate its application in SQL.
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply content/issue-2/write-a-sql-optimizer-using-egg.md:34]: Actually they are S-expressions. But I want to mention Lisp as people are more familiar with it. Would it be better if changed to "Lisp style S-expressions"?

[the remark this answers is not in the document]
Comment (wangrunji0408) [reply content/issue-2/write-a-sql-optimizer-using-egg.md]: Sounds reasonable. I have renamed them.

File: content/issue-2/write-a-sql-optimizer-using-egg.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+For example, the multiplication-to-shift rule mentioned at the beginning can be described in Egg using the [`rewrite`](https://docs.rs/egg/0.9.3/egg/macro.rewrite.html) macro as follows:
+
+```rust
+rewrite!("mul2-to-shl1"; "(* ?a 2)" => "(<< ?a 1)")
+```
+
+The strings on either side of `=>` represent the expression patterns before and after the transformation, where `?a` represents a variable that can be replaced with any expression. When the runner finds an expression that matches this pattern, it generates a new expression according to the pattern on the right and unions it with the original node to form an equivalence class. By the way, `=>` can also be replaced with `<=>` to indicate a bidirectional transformation.
+
+For some rules, they require certain conditions to be met before execution. For example, the following rule is only valid when the divisor is not zero:
```
Comment (mfrw) [inline content/issue-2/write-a-sql-optimizer-using-egg.md:138]: Nit-Pick suggestion -- but probably, I think given the language used so far, it seems you are talking to me ... and I am enjoying the conversation :)
```suggestion
Some rules require certain conditions to be met before execution. For example, the following rule is only valid when the divisor is not zero:
```

**Review by mfrw — APPROVED** [review]
LGTM for english version.
🥇 

Thanks @wangrunji0408

File: content/issue-2/write-a-sql-optimizer-using-egg.md (hunk 4 above)
Comment (wangrunji0408) [reply content/issue-2/write-a-sql-optimizer-using-egg.md]: Thanks!

**Review by Folyd — APPROVED** [review]
Great article, thanks @wangrunji0408.

**Pull request merged**